### ISSUE TITLE
Locking bug in the ConcurrentDictionary implementation?

### ISSUE THREAD

**Issue by paulrobinmanson** [peterprokop/SwiftConcurrentCollections#1]
I believe there's a bug in the implementation of **ConcurrentDictionary**:

```
public subscript(key: Key) -> Value? {
   get {
      return value(forKey: key)
   }
   set {
      rwlock.writeLock()
      guard let newValue = newValue else {
          _remove(key)
          return
      }
      _set(value: newValue, forKey: key)
      rwlock.unlock()
   }
}
```

If the new value being assigned to a specific dictionary key happens to be nil, then rwlock.writeLock() is called but the setter returns without calling rwlock.unlock(). So the **ConcurrentDictionary** will remain locked and all subsequent access attempts will block (forever). 

A solution is to add a defer{ } block which will be guaranteed to call rwlock.unlock() no matter how the setter exits, as follows:

```
public subscript(key: Key) -> Value? {
    get {
       return value(forKey: key)
    }
    set {
        rwlock.writeLock()
        defer {
            rwlock.unlock()
        }
        guard let newValue = newValue else {
            _remove(key)
            return
        }
        _set(value: newValue, forKey: key)
    }
}
```

**Comment by peterprokop** [comment]
Hi @paulrobinmanson 
This is really good spot, let me fix this quickly

**Comment by peterprokop** [comment]
Fixed in v1.2.1
Let me know if you have any other comments

**Comment by paulrobinmanson** [comment]
Thanks Peter. And thanks for publishing the SwiftConcurrentCollections framework. I’ve used a few GCD-based thread-safe Array/Dictionary frameworks in the past and always ran into deadlock issues (primarily due to accessing these objects from the “wrong” queue). The lock-based ConcurrentDictionary and ConcurrentArray seem to work much more reliably in practice (and with less CPU overhead too). So thanks for these!

Cheers,
Paul
